### ISSUE TITLE
HTTPClientFactory sets followRedirect on the HTTPPageGetter class

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3192]
|<img alt="amcnabb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| amcnabb reported|
|-|-|
|Trac ID|trac#3192|
|Type|defect|
|Created|2008-04-18 16:41:59Z|
|Branch|https://github.com/twisted/twisted/tree/follow-redirect-clash-3192|

HTTPClientFactory sets followRedirect on the HTTPPageGetter class rather than on an HTTPPageGetter instance.

HTTPClientFactory sets `self.protocol.followRedirect = followRedirect` in its __init__ method.  Since `self.protocol` is the HTTPPageGetter class, it appears that the client factory is setting the followRedirect attribute for all connections (associated with all HTTPClientFactories).  I was surprised that HTTPClientFactory wouldn't set `self.followRedirect = followRedirect`, in which case HTTPClient would look at `self.factory.followRedirect`.

Thanks.

Attachments:

* [isolated-follow-redirect-3192.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0a3/0a354700948dcf53ff2411801b21bb438b959710/3a88bf16da18b97c4a453603318ba263bceca507.patch) (1962 bytes) - added by jonathanj on 2008-11-27 10:51:01Z - 
* [isolated-follow-redirect-3192-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0a3/0a354700948dcf53ff2411801b21bb438b959710/3790d751f52fd2dd3ddc3e3cb7627c3fac8bce91.patch) (2007 bytes) - added by jonathanj on 2008-11-27 13:40:52Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3192 3192
type__defect defect
reporter__amcnabb amcnabb
priority__normal normal
milestone__ 
branch__branches_follow_redirect_clash_3192 branches/follow-redirect-clash-3192
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__easy easy
time__1208536919000000 1208536919000000
changetime__1227796245000000 1227796245000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

[#3195](https://github.com/twisted/twisted/issues/3195) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)|@jonathanj set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @jonathanj|
|-|-|

For backward compatibility, the protocol should still have and respect a `followRedirect` attribute on itself.

Can you put the attribute there in buildProtocol?  That should preserve the fix and backward compatibility.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)|@jonathanj set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Great.  I think `downloadPage` breaks if `followRedirect` doesn't get a default value on the class, in case someone is using this class with a different factory which doesn't set the attribute on buildProtocol.  I'll drop that part of the diff and then apply.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [25504]) Branching to 'follow-redirect-clash-3192'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [25505]) Apply isolated-follow-redirect-3192-2.patch

refs [#3192](https://github.com/twisted/twisted/issues/3192)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [25506]) re-add HTTPPageGetter followRedirect class attribute

refs [#3192](https://github.com/twisted/twisted/issues/3192)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [25507]) Update docstring

refs [#3192](https://github.com/twisted/twisted/issues/3192)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [25508]) Merge follow-redirect-clash-3192

Author: jonathanj
Reviewer: exarkun
Fixes: [#3192](https://github.com/twisted/twisted/issues/3192)

Fix `getPage` so that the value supplied for `followRedirect` does
not override the value supplied for all previously issued and still
running `getPage` calls.

Previously `HTTPClientFactory` would set the `followRedirect` class
attribute of `HTTPPageGetter` rather than setting it on the one
instance created to satisfy one `getPage` call.  Now it sets the
attribute on an instance instead.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
